torch/csrc/jit/codegen/cuda/lower_index_hoist.cpp

```diff
@@ -12,419 +12,275 @@ namespace cuda {
-      {id}, {td->domain().begin(), td->domain().end()});
-
-  std::unordered_set<IterDomain*> used_leaf_ids;
-
-  for (const auto leaf : td->domain()) {
-    if (std::find(all_vals_between.begin(), all_vals_between.end(), leaf) !=
-        all_vals_between.end()) {
-      used_leaf_ids.insert(leaf);
+// Given a set of loop variables (`dependency`), find the outer-most loop nest
+// that contains all these variables.
+int64_t findLowestLevelWithSatisfiedDependency(
+    const std::vector<kir::ForLoop*>& loops,
+    const std::unordered_set<Val*>& dependency) {
+  int64_t level;
+  for (level = loops.size(); level > 0; level--) {
+    auto loop = loops[level - 1];
+    if (loop->isTrivial()) {
+      continue;
+    }
+    if (dependency.count(loop->index())) {
+      break;
     }
   }
-
-  TORCH_INTERNAL_ASSERT(
-      !used_leaf_ids.empty(),
-      "No used id found: ",
-      id->toString(),
-      ", ",
-      td->toString());
-
-  return used_leaf_ids;
+  return level;
 }
 
-} // namespace
-
-CommonIndexKey::CommonIndexKey(
-    IterDomain* consumer_indexed_id,
-    TensorDomain* consumer_td,
-    TensorDomain* ref_td,
-    const std::unordered_map<IterDomain*, Val*>& ref_index_map,
-    const std::vector<kir::ForLoop*>& loops) {
-  auto gpu_lower = GpuLower::current();
-
-  concrete_indexed_id_ = gpu_lower->caMap()->getConcreteMappedID(
-      consumer_indexed_id, IdMappingMode::EXACT);
-
-  const auto consumer_leaf_ids =
-      getUsedLeafIds(consumer_indexed_id, consumer_td);
-
-  // Convert to Parallel concrete IDs to find matching loops.
-  std::unordered_set<IterDomain*> concrete_leaf_ids;
-  for (auto& id : consumer_leaf_ids) {
-    concrete_leaf_ids.insert(
-        gpu_lower->caMap()->getConcreteMappedID(id, IdMappingMode::LOOP));
+// Test if there is a non-trivial loop between the current loop nests and the
+// given level.
+bool isHoisted(const std::vector<kir::ForLoop*>& loops, int64_t level) {
+  while (level < loops.size()) {
+    auto loop = loops[level++];
+    if (!loop->isTrivial()) {
+      return true;
+    }
   }
+  return false;
+}
 
-  // Find used loops and their index vals
-  for (const auto i : c10::irange(loops.size())) {
-    auto loop = loops.at(i);
-    auto loop_id = gpu_lower->caMap()->getConcreteMappedID(
-        loop->iter_domain(), IdMappingMode::LOOP);
-    auto it = concrete_leaf_ids.find(loop_id);
-    if (it != concrete_leaf_ids.end()) {
-      // This leaf reference id is used for indexing the consumer id
-      used_loops_.push_back(loop);
-      auto index_it = ref_index_map.find(ref_td->axis(i));
-      TORCH_INTERNAL_ASSERT(
-          index_it != ref_index_map.end(),
-          "Index not found for leaf ID, ",
-          ref_td->axis(i)->toString());
-      loop_index_vals_.push_back(index_it->second);
+// Check if in the definition of from, there is a subexpression equivalent to
+// reference. If found, then return this subexpression.
+Val* findCommonSubexpr(Val* from, Val* reference) {
```

Comment:
nit: "common subexpression" would imply it could be a subexpression of both vals, but that's not the case. It's also clear from the comment, but maybe stating it more explicitly would be helpful.

Reply:
Renamed to `findRefAsSubexprOf`